keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+
+As an application developer, I want my application pods not to co-exist with
+specific pods (via PodAntiAffinity). But in some cases it'd be favorable to
+tolerate "violating" pods in a manageable way. For example, suppose an app
+(replicas=2) is using PodAntiAffinity and deployed onto a 2-nodes cluster, and
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+Along with this feature, additional inevitable cost will be applied for each
+cycle of pod scheduling. So to mitigate potential performance impact, initial
+implementation will _probably_ limit the semantics of "even spreading" on
+`kubernetes.io/hostname`.
+
+We also need to make sure that our implementation will not have any performance
+penalty for pods that do not use this feature.
+
+## Design Details
+
+Basically there are two ways in API design and implementation:
+
+1. implemented as a "sub feature" inside Affinity and hence new API applies to
+   `pod.spec.affinity`. (referred as [Design A](#design-b) in this doc)
+1. implemented as a standalone feature and new API applies to `pod.spec`.
+   (referred as [Design B](#design-b) in this doc)
+
+### Design A
+
+A new structure called `EvenSpreading` is introduced and it's only effective
+when it's not nil.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+}
+```
+
+In this design, `EvenSpreading` is a field in Affinity specs:
+
+```go
+type NodeAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAntiAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+### Design B
+
+Unlike Design A, `EvenSpreading` acts as a standalone spec and applies to
+`pod.spec`. Similarly it's only effective when it's not nil.
+
+```go
+type PodSpec struct {
+    EvenSpreading *EvenSpreading
```

Comment:
i like this better :-)

Reply:
Yeap, this is our preference for now.

Reply:
where do we say hard or soft ? or is that controlled by maxSkew ?

Reply:
It's controlled similarly like what we did in PodAffinity (via podAffinityTerms), not by maxSkew. But to be honest, I'm pretty a lot of work  to implement both versions in one release :)